build.gradle.kts

```diff
@@ -0,0 +1,143 @@
+}
+
+// Expose DiSCo version to subprojects
+val discoVersion by extra("0.10.0")
+
+subprojects {
+    version = "2.7.0"
+    group = "com.amazonaws"
+
+    repositories {
+        mavenCentral()
+        mavenLocal()
+    }
+
+    // Configure the shadow jar task, which does shading, to run after Gradle runs the jar task
+    pluginManager.withPlugin("com.github.johnrengelman.shadow") {
+        tasks {
+            named<DefaultTask>("assemble") {
+                dependsOn(named<ShadowJar>("shadowJar"))
+            }
+            named<ShadowJar>("shadowJar") {
+                // Suppress the "-all" suffix on the jar name, simply replace the default built jar instead
+                archiveClassifier.set(null as String?)
+
+                dependsOn(named<Jar>("jar"))
+            }
+        }
+    }
+
+    plugins.withId("java") {
+        configure<JavaPluginExtension> {
+            sourceCompatibility = JavaVersion.VERSION_1_8
+            targetCompatibility = JavaVersion.VERSION_1_8
+
+            withJavadocJar()
+            withSourcesJar()
+        }
+
+        dependencies {
+            // BOMs for common projects
+            add("implementation", platform("com.amazonaws:aws-xray-recorder-sdk-bom:2.7.0"))
+            add("implementation", platform("software.amazon.disco:disco-toolkit-bom:${discoVersion}"))
+            add("implementation", platform("com.fasterxml.jackson:jackson-bom:2.11.0"))
+            add("implementation", platform("com.amazonaws:aws-java-sdk-bom:1.11.837"))
+            add("implementation", platform("software.amazon.awssdk:bom:2.13.70"))
+
+            // TODO: Add build step for running Null checker
+            add("compileOnly", "org.checkerframework:checker-qual:3.4.1")
+
+            // Common test dependencies
+            add("testImplementation", "junit:junit:4.12")
+            add("testImplementation", "org.assertj:assertj-core:3.16.1")
+            add("testImplementation", "org.mockito:mockito-core:2.28.2")
+        }
+    }
+
+    plugins.withId("maven-publish") {
+        plugins.apply("signing")
+
+        // Disable publishing a bunch of unnecessary Gradle metadata files
+        tasks.withType<GenerateModuleMetadata> {
+            enabled = false
+        }
+
+        // Defer maven publish until the assemble task has finished, giving time for shadowJar to complete if present
+        tasks.withType<AbstractPublishToMaven> {
+            dependsOn(tasks.named<DefaultTask>("assemble"))
+        }
+
+        configure<PublishingExtension> {
+            publications {
+                register<MavenPublication>("maven") {
+
+                    // If the shadow plugin is present, we should publish the shaded artifact
+                    // Otherwise, just publish the standard JAR
+                    plugins.withId("java") {
+                        if (plugins.hasPlugin("com.github.johnrengelman.shadow")) {
+                            artifact(tasks.named<Jar>("jar").get())
```

Comment:
Is this the shaded artifact?

Reply:
Yes, although I think this actually might have to be wrapped in `afterExecute` for the `hasPlugin` to work. I'll add that.

Reply:
Was there something you noticed awry?

Reply:
Ah sorry should have been explicit - it's referring to "jar" task, not "shadowJar" task

Reply:
Instead of afterEvaluate, you can just use `plugins.WithId("com.github.johnrengelman.shadow")`

Reply:
Yeah I thought about this but I couldn't figure out a way to get gradle to behave differently if a plugin wasn't present. E.g. something like:

```java
plugins.withId("shadow") {
  artifact()
}

plugins.withoutId("shadow") {
  from(components["java"])
}
```

The `afterEvaluate` + `hasPlugin` approach worked, but let me know if there's a preferable `withId` way instead.